DOI-USGS/snow-to-flow
# Update node

Changes made:
-----------
This PR updates our npm configuration to use `node` version 20 and re-point the docker build to `debian:stable`,  because `debian:stretch` has been deprecated.

### To test
1. Get off VPN if you are on VPN
2. Check what version of node you are using by running `node -v`
3. If not using version 20, switch to node version 20 by running `nvm use 20`  _[If it's not installed, install latest version of node by running `nvm install 20.0.0 --reinstall-packages-from=14.17.0`]_
4. Delete `package-lock.json` and `node_modules` directory
5. Run `npm install`
6. Run `npm run serve`

### Changes:
* In `Dockerfile`, point to to a stable debian release, instead of the archived stretch release, using the debian:stable tag listed as a current tag [on Dockerhub](https://hub.docker.com/_/debian/tags).
* In `Dockerfile`, get and install the current version of node ([version 20](https://github.com/nodejs/release#release-schedule))
* In package.json remove addition of `node-sass` in `devDependencies` and instead specify latest version of `sass`, as [`node-sass` is not supported for node 20](https://github.com/sass/node-sass#node-version-support-policy). `Node-sass` is now [deprecated](https://github.com/sass/node-sass), and the makers recommend switching to [`sass`](https://sass-lang.com/dart-sass)
* In `package.json` add `env NODE_OPTIONS=--openssl-legacy-provider` to all 'serve' and 'build' `scripts` to overcome a lack of support for digital envelope routines in `node` 17+. See [this issue](https://github.com/webpack/webpack/issues/14532) and [this comment](https://github.com/webpack/webpack/issues/14532#issuecomment-949699325)
* Update other miscellaneous packages to match versions used in more recently built and updated `package.json` in [what-is-drought](https://github.com/USGS-VIZLAB/what-is-drought/blob/main/package.json)
* Add `.npmrc` file in root directory with the content legacy-peer-deps=true to overcome peer dependency conflicts between `cli-plugin-eslint`, `eslint`, and `babel-eslint`. See [this issue](https://github.com/vuejs/vue-cli/issues/7149) and [this comment](https://github.com/vuejs/vue-cli/issues/7149#issuecomment-1221103008).
* Changes to other files due to running lint:
![image](https://github.com/DOI-USGS/snow-to-flow/assets/54007288/e6e428e7-5ab7-421e-897f-52054e6fb5fd)


Testing:
----------------------------
Before making this pull request, I:

I can confirm this has been checked on:
